File: 05-Hash Tables/cpp/main.cpp
#include <iostream>
#include <unordered_map>
#include <set>
#include <functional>

#include"CharFinder.h"

using namespace std;

int Hash(string key);

int main() {

//    Built-in HashMap
    cout << "Built-in HashMap" << endl;

    unordered_map<int, string> myMap;
    myMap[1] = "Mosh";
    myMap[2] = "John";
    myMap[3] = "Mary";

    for (const auto &pair: myMap) {
        cout << pair.first << ", " << pair.second << endl;
    }

    cout << "--------" << endl;

//    findFirstNonRepeatingChar
    cout << "findFirstNonRepeatingChar" << endl;

    CharFinder charFinder;
    cout << charFinder.findFirstNonRepeatingChar("A Green Apple") << endl;
    cout << charFinder.findFirstNonRepeatingChar("An Apple is Green") << endl;

    cout << "--------" << endl;

//    Built-in set
    cout << "Built-in set" << endl;

    set<int> s;
    int numbers[] = {1, 2, 3, 3, 2, 1, 4};
    for (auto number: numbers) {
        s.insert(number);
    }
    for (auto x: s) {
        cout << x << " ";
    }
    cout << endl;

    cout << "--------" << endl;

//    findFirstRepeatedChar
    cout << "findFirstRepeatedChar" << endl;

    cout << charFinder.findFirstRepeatedChar("A Green Apple") << endl;
    cout << charFinder.findFirstRepeatedChar("An Apple is Green") << endl;

    cout << "--------" << endl;

//    Hash Functions
    cout << "Hash Functions" << endl;

    cout << Hash("123456-A") << endl;

    hash<string> str;
    cout << str("123456-A") << endl;

    cout << "--------" << endl;


    return 0;
}

int Hash(string key) {
    int hash = 0;
    for (auto ch: key)
        hash += ch;


    return hash % 100;
}

File: 05-Hash Tables/java/src/Main.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Main {
    public static void main(String[] args) {
//        Built-in HashMap
        System.out.println("Built-in HashMap");

        Map<Integer, String> map = new HashMap<>();
        map.put(1, "Mosh");
        map.put(2, "John");
        map.put(3, "Mary");

//        overwrite the duplicate key
        map.put(3, "Marianne");

//        null values is allowed
        map.put(4, null);

//        null key is also allowed
        map.put(null, null);

        System.out.println(map);

        System.out.println(map.containsKey(3));          //O(1)
        System.out.println(map.containsValue("Mosh"));   //O(n)

        for (var item : map.entrySet())
            System.out.println(item);

        System.out.println("--------");

//        findFirstNonRepeatingChar
        System.out.println("findFirstNonRepeatingChar");

        CharFinder charFinder = new CharFinder();
        System.out.println(charFinder.findFirstNonRepeatingChar("A Green Apple"));
        System.out.println(charFinder.findFirstNonRepeatingChar("An Apple is Green"));

        System.out.println("--------");

//        Built-in HashSet
        System.out.println("Built-in HashSet");

        Set<Integer> set = new HashSet<>();

        int[] numbers = {1, 2, 3, 3, 2, 1, 4};
        for (var number : numbers)
            set.add(number);
        System.out.println(set);

        System.out.println("--------");

//        findFirstRepeatedChar
        System.out.println("findFirstRepeatedChar");

        System.out.println(charFinder.findFirstRepeatedChar("green apple"));
        System.out.println(charFinder.findFirstRepeatedChar("An Apple is Green"));

        System.out.println("--------");

//        Hash Functions
        System.out.println("Hash Functions");

        Map<String, String> newMap = new HashMap<>();
        newMap.put("123456-A", "Mosh");
        System.out.println(hash("123456-A"));

        String str = "orange";
        System.out.println(str.hashCode());

        System.out.println("--------");

//        Custom Hash Table
        System.out.println("Custom Hash Table");

        HashTable table = new HashTable(5);
        table.put(6, "A");
        table.put(8, "B");
        table.put(11, "C");
        table.put(6, "A+");
//        table.put(6, "null");
        table.remove(6);
        System.out.println(table.get(6));

        System.out.println("--------");

    }

    public static int hash(String key) {
        int hash = 0;
        for (var ch : key.toCharArray())
            hash += ch;

        return hash % 100;
    }
}